doc/icastats.1
.\" icastats man page source
.\"
.\" use
.\"   groff -man -Tutf8 icastats.1
.\" or
.\"   nroff -man icastats.1
.\" to process this source
.\"
.TH ICASTATS 1 2013-12-06 IBM "icaststats user manual"
.SH NAME
icastats \- display statistic data for the libica cryptographic functions
.SH SYNOPSIS
.B icastats
[-v | --version] [-h | --help] [--reset-all | -R] [--reset | -r]
[--delete-all |-D] [--delete | -d] [--all | -A] [--summary | -S] [-U |
--user <username>] [--key-sizes | -k]
.SH DESCRIPTION
.B icastats
displays statistic data about the usage of cryptographic functions provided by
libica.
.br
Libica is a cryptographic library supporting SHA, RSA, ECC, DES and AES in
different modes of operations. The invocation of each call to all the
cryptographic functions is tracked with individual counters which can be
displayed and maintained with icastats.
.br
Here is a shortened sample output:
.P
.nf
 function     |       # hardware      |      # software
--------------+-----------------------+-------------------------
              |     ENC   CRYPT   DEC |      ENC  CRYPT    DEC
--------------|-----------------------|-------------------------
        SHA-1 |             0         |             0
      SHA-224 |             0         |             0
      SHA-256 |             0         |             0
          ...
   ECDSA Sign |             0         |             0
 ECDSA Verify |             0         |             0
    EC Keygen |             0         |             0
       RSA-ME |             0         |             0
          ...
       RSA-ME |             0         |             0
      RSA-CRT |             0         |             0
          ...
      AES CFB |       0             0 |       0            0
      AES CTR |       0             0 |       0            0
     AES CMAC |       0             0 |       0            0
      AES XTS |       0             0 |       0            0
      AES GCM |       0             0 |       0            0
.fi
.P
For each cryptographic function the table shows the number of invocations
where hardware support was involved or the libica fall back implementation
in software was used. For the ciphering methods the invocation counter is
further divided into encrypt and decrypt operation counter values.
For some cryptographic functions (AES, RSA, ECC) counters exist per key size.
Use the --key-sizes | -k option to display those counters as well.
.P
All the counter values are stored and maintained in one shared memory page
for each user. This memory area is created automatically with the first run
of icastats and persists until it is explicitly removed (see the -d option)
or system shut down. This also means that the statistical data shown with
icastats is on a per user base and only the root user is able to see and
maintain the collection of statistic data from libica at system scope.
.P
Newer operating systems might remove these shared memory segments
after the user has logged out from the system (systemd cleanup action).
In this case all collected statistic data are gone.
To prevent the statistic data from being removed you can configure the
systemd to keep all shared memory segments after session exit by
enabling the "RemoveIPC=no" paramater in "/etc/systemd/logind.conf".
Alternatively you can setup the systemd user manager to enable user
lingering by typing "loginctl enable-linger <user>".

Note that one single libica function may increase several different counters
when internally using different hardware functions. For example, performing
AES GCM on a z13 involves using the AES ECB, AES CTR and GHASH hardware
functions. On a z14, the AES GCM counter increases to indicate the use of the
KMA instruction. Depending on the input data, other counters may also increase.
Therefore, by looking at the hardware counters, it is not possible to see
how often a particular API function was called.

.SH OPTIONS
.IP "-v or --version"
show libica version and copyright
.IP "-h or --help"
display this help and exit
.IP "-R or --reset-all"
reset the statistic data for each user that has a shared memory segment to
zero (only root user)
.IP "-r or --reset"
reset the statistic data for the current user to zero
.IP "-D or --delete-all"
delete all shared memory segments from all users who have an icastats shared
memory segment (only root user)
.IP "-d or --delete"
delete the libica shared memory area for the current user
.IP "-A or --all"
show libica statistic data for each user in the system where a shared
memory area exists in multiple tables (only root user)
.IP "-S or --summary"
show libica statistic data for each user in the system where a shared
memory area exists in one table accumulating all the per user counter
values (only root user)
.IP "-U <username> or --user <username>"
show statistic values from the given user (only root user)
.IP "-k or --key-sizes"
show libica statistic data per key size
.SH FILES
.nf
/shm/dev/icastats_<userid>
.fi
.SH RETURN VALUE
.IP 1
unsupported or unknown argument, argument which requires root privileges
given but not root user, failures at creating or attaching to the shared
memory area ...
.IP 0
normal and sucessful program execution
.SH SEE ALSO
icainfo (1)


ChangeLog
v3.9.0
 - [FEATURE] Add support for OpenSSL 3.0
   [FEATURE] icainfo: new parm -c to display available EC curves
v3.8.0
 - [FEATURE] provide libica-cex module to satisfy special security requirements
   [FEATURE] FIPS: enforce the HMAC check
 - [UPDATE] exploit autoselect domain support for ECC
 - [UPDATE] FIPS: Add SHA3 kats to fips powerup tests
v3.7.0
 - [FEATURE] FIPS: Add HMAC based library integrity check
 - [PATCH] icainfo: bugfix for RSA and EC related info for software column.
 - [PATCH] FIPS: provide output iv in cbc-cs decrypt as required by FIPS tests
 - [PATCH] FIPS: Fix DES and TDES key length
 - [PATCH] icastats: Fix stats counter format
v3.6.1
 - [PATCH] Fix x25519 and x448 handling of non-canonical values
v3.6.0
 - [FEATURE] Add MSA9 CPACF support for Ed25519, Ed448, X25519 and X448
v3.5.0
 - [FEATURE] Add MSA9 CPACF support for ECDSA sign/verify
v3.4.0
 - [FEATURE] Add SHA-512/224 and SHA-512/256 support
v3.3.3
 - [PATCH] Various bug fixes
v3.3.2
 - [PATCH] Skip ECC tests if required HW is not available
 - [PATCH] Update spec file
v3.3.1
 - [PATCH] Fix configure.ac to honour CFLAGS
v3.3.0
 - [FEATURE] Add CEX supported elliptic-curve crypto interfaces
 - [FEATURE] Add SIMD supported multiple-precision arithmetic interfaces
 - [FEATURE] Add interface to enable/disable SW fallbacks
 - [FEATURE] Add 'make check' target, test-suite rework
v3.2.1
 - [FEATURE] Use z14 PRNO-TRNG to seed SHA512-DRBG.
 - [PATCH] Various bug fixes.
v3.2.0
 - [FEATURE] New AES-GCM interface.
 - [UPDATE] Add symbol versioning.
v3.1.1
 - [PATCH] Various bug fixes related to old and new AES-GCM implementations.
 - [UPDATE] Add SHA3 test cases. Improved and extended test suite.
v3.1.0
 - [FEATURE] Add KMA support for AES-GCM.
 - [FEATURE] Add SHA-3 support.
 - [PATCH] Reject RSA keys with invalid key-length.
 - [PATCH] Allow zero output length for ica_random_number_generate.
 - [PATCH] icastats: Correct owner of shared segment when root creates it.
v3.0.2
 - [PATCH] Fix locking callbacks for openSSL APIs.
v3.0.1
 - [PATCH] Fixed msa level detection on zEC/BC12 GA1 and predecessors.
v3.0.0
 - [FEATURE] Added FIPS mode.
 - [PATCH] Sanitized exported symbols.
 - [PATCH] Removed deprecated APIs. Marked some APIs as deprecated.
 - [PATCH] Adapted to OpenSSL v1.1.0.
 - [PATCH] RSA key generation is thread-save now.
v2.6.2
 - [PATCH] Performance improvements for des-ctr, 3des-ctr and aes-ctr
v2.6.1
 - [Patch] Fixed buffer overflow on random generation
 - [Patch] Fixed DRBG entropy input
 - [Patch] Fixed icastats test case
v2.6.0
 - [FEATURE] Added NIST compliant deterministic random bit generator (DRBG)
v2.5.0
 - [FEATURE] Added streaming support for Galois Counter Mode (AES-GCM)
 - [FEATURE] Allow RSA Key generation for any granularity (57..4096 bits)
 - [FEATURE] New RSA CRT Key check API (Convertion of non-compliant keys)
 - [PATCH] Replaced deprecated RSA_key_generate function
 - [PATCH] Fixed accounting for GCM, CCM and CMAC. Added accounting for GHASH
 - [PATCH] Improved configure script to automatically build test cases
 - [UPDATE] Added SHA test cases. Improved and extended test suite
 v2.4.0
 Reworked the statistic measurement facility (icastats)
 - Statistics are now collected independent from the process context
 - Statistic data gathered per user
 - User based statistic management (reset/delete/summerize)
v2.3.0
 Changes since Version 2.2.0:
 - Introduced new API to recieve list of supported algorithms
v2.2.0
 Introduce support for more MSA4 extensions.
 New symetric encryption modes:
 - CBC-CS (DES, TDES, AES-[128|192|256])
 - CMAC (DES, TDES, AES-[128|192|256])
 - CCM (AES-[128|192|256])
 - GCM (AES-[128|256])
v2.1.1
 Changes since Version 2.1.0:
 - [PATCH] synchronize shared memory ref counting
 - [PATCH] Fix result/error handling in testcase for ica_get_version()
 - [PATCH] Fix temporary buffer allocation in ica_get_version()
 - [UPDATE] Interface (API) description
 - [PATCH] Fix commandline argument checking in RSA testcase
 - [PATCH] Remove duplicate entry from target list
v2.1.0
- Introduce support for MSA4 extensions. New symmetric encryption modes:
  CFB, CTR, OFB, XTS, CMAC
- New API calls:
  - ica_get_version(...)
v2.0.6
Fix for tracker item 3202845.
  Improve online help for icastats.
v2.0.5
Fix for performance improvement
  Remove signal handler wrapping of crypto instructions.
v2.0.4
Fix for STCK buffer length
  Correctly adjust the STCK buffer length in the PRNG based on whether
  we are running in 31 or 64 bit.
v2.0.3
minor wording fix in icastats tool:
  changed output wording from SHAxxx to SHA-xxx
Bugfix for:
  libica abends with illegal instuction when CPACF is disabled.
v2.0.2
Bug fixes in asm, stcke for 31 bit, and fixes in programming samples
v2.0.1
Add locking for shared memory segment, which is needed when icastats
is linked dynamically against libica.
v2.0
Substitute software fallbacks with lowlevel OpenSSL function calls.
Rewrite most of libica.
Only Linux on System z code.


src/icastats.c
/* This program is released under the Common Public License V1.0
 *
 * You should have received a copy of Common Public License V1.0 along with
 * with this program.
 */

/*
 * Authors(s): Christian Maaser <cmaaser@de.ibm.com>
 *             Holger Dengler <hd@linux.vnet.ibm.com>
 *             Benedikt Klotz <benedikt.klotz@de.ibm.com>
 *             Ingo Tuchscherer <ingo.tuchscherer@de.ibm.com>
 *
 * Copyright IBM Corp. 2009-2019
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#include <pwd.h>
#include <libgen.h>
#include "icastats.h"

#define CMD_NAME "icastats"
#define COPYRIGHT "Copyright IBM Corp. 2009-2019"

void print_version(void)
{
	printf(CMD_NAME ": libica version " VERSION "\n" COPYRIGHT "\n");
}

void print_help(char *cmd)
{
	printf("Usage: %s [OPTION]\n\n", cmd);
	printf("This command is used to indicate whether libica uses hardware crypto functions or\n"
	       "software fallbacks. It provides an overview of the algorithms with modes of operation.\n"
	       "\n"
	       "Options:\n"
	       " -r, --reset         set the own function counters to zero.\n"
	       " -R, --reset-all     reset the statistsics from all users. (root user only)\n"
	       " -d, --delete        delete your own statistics.\n"
	       " -D, --delete-all    delete the statistics from all users. (root user only)\n"
	       " -U, --user <userid> show the statistics from one user. (root user only)\n"
	       " -S, --summary       show the accumulated statistics from alle users. (root user only)\n"
	       " -A, --all	     show the statistic tables from all users. (root user only)\n"
	       " -k, --key-sizes     show statistics per key size.\n"
	       " -v, --version       output version information\n"
	       " -h, --help          display help information\n");
}

#define getopt_string "rRdDU:SAkvh"
static struct option getopt_long_options[] = {
	{"reset", 0, 0, 'r'},
	{"reset-all", 0, 0, 'R'},
	{"delete", 0, 0, 'd'},
	{"delete-all", 0, 0, 'D'},
	{"user", required_argument, 0, 'U'},
	{"summary", 0, 0, 'S'},
	{"all", 0, 0, 'A'},
	{"key-sizes", 0, 0, 'k'},
	{"version", 0, 0, 'v'},
	{"help", 0, 0, 'h'},
	{0, 0, 0, 0}
};

const char *const STATS_DESC[ICA_NUM_STATS] = {
	STAT_STRINGS
};



#define CELL_SIZE 12
void print_stats(stats_entry_t *stats, int key_sizes)
{
	printf(" function       |             hardware         |              software\n");
	printf("----------------+------------------------------+-----------------------------\n");
	printf("                |        ENC    CRYPT     DEC  |        ENC     CRYPT    DEC \n");
	printf("----------------+------------------------------+-----------------------------\n");
	unsigned int i;
	for (i = 0; i < ICA_NUM_STATS; ++i) {
		if (!key_sizes && strncmp(STATS_DESC[i], "- ", 2) == 0)
			continue;

		if(i <= ICA_STATS_RSA_CRT_4096) {
			printf(" %14s |        %*lu          |         %*lu\n",
			       STATS_DESC[i],
			       CELL_SIZE,
			       stats[i].enc.hw,
			       CELL_SIZE,
			       stats[i].enc.sw);
		} else {
			printf(" %14s |%*lu     %*lu |%*lu    %*lu\n",
			       STATS_DESC[i],
			       CELL_SIZE,
			       stats[i].enc.hw,
			       CELL_SIZE,
			       stats[i].dec.hw,
			       CELL_SIZE,
			       stats[i].enc.sw,
			       CELL_SIZE,
			       stats[i].dec.sw);

	       }
	}
}




int main(int argc, char *argv[])
{
	int rc = 0;
	int index = 0;
	int reset = 0;
	int delete = 0;
	int sum = 0;
	int user = -1;
	int all = 0;
	int key_sizes = 0;
	struct passwd *pswd;

	while ((rc = getopt_long(argc, argv, getopt_string,
				 getopt_long_options, &index)) != -1) {
		switch (rc) {
		case 'r':
			reset = 1;
			break;
		case 'R':
			if(geteuid() != 0){
				fprintf(stderr,"You have no rights to reset all shared memory"
					" segments!\n");
				return EXIT_FAILURE;
			}
			reset = 2;
			break;
		case 'd':
			delete = 1;
			break;
		case 'D':
			if(geteuid() != 0){
				fprintf(stderr,"You have no rights to delete all shared memory"
					" segments!\n");
				return EXIT_FAILURE;
			}

			delete = 2;
			break;
		case 'U':
			if((pswd = getpwnam(optarg)) == NULL){
				fprintf(stderr, "The username %s is not known"
					" on this system.\n", optarg );
				return EXIT_FAILURE;
			}
			user = pswd->pw_uid;
			break;
		case 'S':
			sum = 1;
			break;
		case 'A':
			all = 1;
			break;
		case 'k':
			key_sizes = 1;
			break;
		case 'v':
			print_version();
			exit(0);
			break;
		case 'h':
			print_help(basename(argv[0]));
			exit(0);
		default:
			fprintf(stderr,
				"Try '%s --help' for more information.\n",
				basename(argv[0]));
			return EXIT_FAILURE;
		}
	}

	if (optind < argc) {
		fprintf(stderr, "%s: invalid option.\n\
			Try '%s --help' for more information.\n",
			argv[0], basename(argv[0]));
		return EXIT_FAILURE;
	}

	if(delete == 2){
		if(delete_all() == -1){
			perror("deleteall: ");
			return EXIT_FAILURE;
		}
		return EXIT_SUCCESS;
	} else if(delete){
		stats_mmap(user);
		stats_munmap(SHM_DESTROY);
		return EXIT_SUCCESS;
	}
	if(all){
		char *usr;
		stats_entry_t *entries;
		while((usr = get_next_usr()) != NULL){
			if((entries = malloc(sizeof(stats_entry_t)*ICA_NUM_STATS)) == NULL){
				perror("malloc: ");
				return EXIT_FAILURE;
			}
			get_stats_data(entries);
			printf("user: %s\n", usr);
			print_stats(entries, key_sizes);
			free(entries);
		}
		return EXIT_SUCCESS;
	}

	if (sum){
		stats_entry_t *entries;
		if((entries = malloc(sizeof(stats_entry_t)*ICA_NUM_STATS)) == NULL){
			perror("malloc: ");
			return EXIT_FAILURE;
		}

		if(!get_stats_sum(entries)){
			perror("get_stats_sum: ");
			return EXIT_FAILURE;
		}
		print_stats(entries, key_sizes);
		return EXIT_SUCCESS;


	}

	if(reset == 2){
		while(get_next_usr() != NULL)
			stats_reset();
		return EXIT_SUCCESS;

	}
	/* Need to open shm before it can be reseted */
	if (stats_mmap(user)) {
		fprintf(stderr, "Could not map shared memory region to local "
			"address space.\n");
		return EXIT_FAILURE;
	}

	if (reset) {
		stats_reset();
	} else{
		stats_entry_t *stats;
		if((stats = malloc(sizeof(stats_entry_t)*ICA_NUM_STATS)) == NULL){
			perror("malloc: ");
			return EXIT_FAILURE;
		}
		get_stats_data(stats);
		print_stats(stats, key_sizes);

	}
	return EXIT_SUCCESS;
}
